functional/lenses.md

```diff
@@ -0,0 +1,481 @@
+    let item = match item.downcast::<CreditRecord>() {
+        Ok(credit) => return Some(credit.customer_id),
+        Err(bx) => bx
+    };
+    let item = match item.downcast::<AccountRecord>() {
+        Ok(acct) => return Some(acct.primary_customer_id),
+        Err(bx) => bx
+    };
+    None
+}
+
+// this is where a lens can be used in a generic way, since "unique ID"
+// need not be the same kind of ID if they are all u64s
+fn unique_ids<I>(iterator: I) -> HashSet<u64>
+    where I: Iterator<Item=Box<dyn Any>>
+{
+    // would be: iterator.filter_map(|it| it.view(get_customer_data)).collect()
+    iterator.map(get_customer_data).filter_map(|uniq_id_opt| uniq_id_opt).collect()
+}
+```
+
+While it would be much cleaner if Rust supported more functional traits (a tall
+order indeed), hopefully the general idea is clear.
+
+## Modification with Lenses
+
+The functional approach need not be restricted to accessing members.
+More powerful lenses can be created which both set and get data in a structure.
+This becomes more interesting when examining how these lenses can be composed
+together.
+
+To begin, consider a new lens trait, this one to update the balance of a bank
+account.
+These accounts have different structures for individuals and commercial
+business customers.
+
+```rust,ignore
+pub struct Eur(u64); // 100 = 1 €
+
+// type casting to integers and display functions omitted
+
+pub trait BalanceChange {
+    fn balance(&self) -> Eur;
+    fn deposit(&mut self, Eur);
+    fn withdraw(&mut self, Eur);
+}
+```
+
+This trait allows *composition* between two lenses.
+It looks very different in Haskell, so here is the Rust way:
+
+```rust,ignore
+use chrono::prelude::*;
+use rand::{RngCore, thread_rng};
+use sha2::{Sha256, Digest};
+
+fn make_id_deposit<A: BalanceChange + CustomerId>(account: &mut A) {
+    // this computes a hash based on the current balance and the account ID,
+    // resulting in a pair of transactions that the customer must use to verify
+    // their identity when resetting their pin via telephone
+
+    let balance = account.balance();
+    let id = account.get_customer_id();
+    let date = Utc::now().timestamp() as usize;
+    let secure: u64 = thread_rng().next_u64();
+
+    let mut hasher = Sha256::new();
+    hasher.update(
+        format!("{:?},{},{},{}", balance, id, date, secure).as_bytes());
+    let cents_hash = &hasher.finalize()[0..1];
+
+    // example: 1234 results in deposits of 0.12 and 0.34
+    let cents = i16::from_be_bytes([cents_hash[0], cents_hash[1]]) as u64 % 1000;
+    account.deposit(Eur(cents / 100));
+    account.deposit(Eur(cents % 100));
+}
+```
+
+Here the composition is done through the generic `A`, which is defined as types
+with both traits.
+In order to literally compose lens functions, Rust uses `map` and `and_then`
+on `Option<T>` and iterator, which would allow each lens to be called
+sequentially on items.
+
+## Prisms: Lenses, but Sum Types
```

Comment:
Not sure whether the sum type part makes sense.

Reply:
Ok, after reading "Functional references: Lens and other Optics in Scala" you reference I now get how you ended up talking about sum-types. But I'd still choose a different term here.

Yes, if you talk about abstract typing the term sum-type is correct. However, it will probably be confusing here because when we talking about sum-types we usually talk about _vales_ being of a sum-type. That's true for enums and trait objects (and a reference to an object implementing some interface in Java), but the very example you choose (serde) is a prime example that prisms may still have value if you don't actually have any run-time variability at all in your final program (i.e. no sum-types are involved, at least not in that specific way).

I'd rather go for "Prisms: higher order Lenses" or something, even if it's also not really correct.